### PR TITLE
AED-766 Gasless NFT minting

### PR DESCRIPTION
**Pull request by arbitrary-nomad**
Added a new template that allow users to deploy a custom gasless ERC1155 and use autotasks to sign and relay minting requests.

### CODE REVIEW

File: defender/gasless-nft-minting/README.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+This section will help you set up the following:
+- Acquire Defender and Relay keys
+- Add funds to the Relay
+- Deploy contracts
+  - `yarn deploy contract --contract-name MinimalForwarder`
+  - `yarn deploy contract --contract-name DemoNFT "0xMINIMAL_FORWARDER_ADDRESS"`
+- Update `config.dev.yml` with the 2 deployed contract addresses
+- Deploy the Autotasks with `serverless deploy`
+- Connect Relays to deployed Autotasks
+  - Contract Deployer relay must be connected to the signing Autotask
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/README.md:16]:
```suggestion
  - Contract Deployer Relay must be connected to the signing Autotask
```

File: defender/gasless-nft-minting/README.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+In the `gasless-nft-minting` directory, perform the following steps:
+
+- A secrets file can be created for each stage of production. We will create one for development:
+  - Copy and rename the `sample.secrets.yml` to `.secrets/dev.yml`
+  - Modify the four lines in the `.secrets/dev.yml` file, replacing the portion in the angle brackets `<>` with your Defender API keys and Secret keys, as indicated:
+  - `defender-api-key: <API Key goes here>`
+  - `defender-api-secret: <Secret key goes here>`
+  - `relay-api-key: <Relay API Key goes here>`
+  - `relay-api-secret: <Relay Secret key goes here>`
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/README.md:59]:
```suggestion
    - `defender-api-key: <API Key goes here>`
    - `defender-api-secret: <Secret key goes here>`
    - `relay-api-key: <Relay API Key goes here>`
    - `relay-api-secret: <Relay Secret key goes here>`
```

File: defender/gasless-nft-minting/README.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+- Click on **Gasless NFT Minting Signer Autotask** Autotask
+- Click on the settings **gear/cog** at the top middle, and select **Edit Settings**
+- In the **Connect a Relayer** section, choose the Relayer that was used to deploy the contract
+
+The same thing must be done for the Autotask that will relay the request:
+
+- Click on the **Autotask** link on the left side
+- Click on **Gasless NFT Minting Relayer Autotask** Autotask
+- Click on the settings **gear/cog** at the top middle, and select **Edit Settings**
+- In the **Connect a Relayer** section, choose ANY Relayer that has enough funds to execute the minting requests on that are submitted to it
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/README.md:104]:
```suggestion
- In the **Connect a Relayer** section, choose ANY Relayer that has enough funds to execute the minting requests that are submitted to it
```

File: defender/gasless-nft-minting/README.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+  - Save the Autotask URL as the value `relayer-webhook` in your `config.dev.yml` file
+
+
+## Typical Usage
+
+The Signer Autotask can now authorize minting of Tokens without spending any gas by signing a minting request and sending it back to the requestor via HTTP.
+
+A user with a signed request can send the request to the Relayer Autotask via HTTP and also not spend gas.
+
+The Relayer Autotask verify the request and signature and execute the transaction on-chain if they are valid.
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/README.md:126]:
```suggestion
The Relayer Autotask verifies the request and signature and executes the transaction on-chain if they are valid.
```

File: defender/gasless-nft-minting/README.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+### Signing and Relaying with scripts
+
+A `sign` and `relay` script are included in this template use the deployed Autotasks.
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/README.md:130]:
```suggestion
The `sign` and `relay` scripts included in this template use the deployed Autotasks. 
```

File: defender/gasless-nft-minting/README.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+### Submitting Requests with Curl
+
+Signature requests can be submitted with curl. Replace the URL with your Signer Autotask webhook URL and 
+```shell
+curl -X POST https://api.defender.openzeppelin.com/autotasks/xxxx/runs/webhook/xxxx \
+   -H 'Content-Type: application/json' \
+   -d '{"address": "0xRECIPIENT_ADDRESS"}'
+```
+Relay requests submitted with curl. Replace the URL with your Relay Autotask webhook URL
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/README.md:144]:
```suggestion
Relay requests can also be submitted with curl. Replace the URL with your Relay Autotask webhook URL
```

File: defender/gasless-nft-minting/README.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+A `sign` and `relay` script are included in this template use the deployed Autotasks. 
+
+In the `gasless-nft-minting` directory, perform the following steps:
+- `yarn sign <address>` - to have the Signer Autotask sign a mint request
+- `yarn relay '<stringified JSON>'` - to relay the request and mint the NFT to the user
+
+### Submitting Requests with Curl
+
+Signature requests can be submitted with curl. Replace the URL with your Signer Autotask webhook URL and
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/README.md:138]:
```suggestion
Signature requests can be submitted with curl. Replace the URL with your Signer Autotask webhook URL and the address with the recipient address:
```

File: defender/gasless-nft-minting/README.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+      "nonce":"0",
+      "to":"0xNFT_ADDRESS",
+      "from":"0xRELAY_ADDRESS","data":"0x731133e9000000000000000000000000USER_ADDRESS00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000"
+    }
+  }'
+```
+
+## Tests
+
+Check that contract can mint with both standard and gasless transactions
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/README.md:162]:
```suggestion
Check that the contract can mint with both standard and gasless transactions
```

File: defender/gasless-nft-minting/README.md [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+- `yarn test tests/nftContract.test.js`
+Check that the configured Defender Relay can sign and verify meta transactions
+- `yarn test tests/eip712sig.online.test.js`
+
+## Notes
+This is a proof of concept template, the Autotask currently allow for unlimited mints of NFTs and can be easily abused until the Relay runs out of funds.
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/README.md:168]:
```suggestion
This is a proof of concept template. The Autotask currently allows for unlimited mints of NFTs and can be easily abused until the Relay runs out of funds.
```

File: defender/gasless-nft-minting/autotasks/relay/index.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
+const stackName = 'gasless_nft_minting';
+const forwarderAddressSecretName = `${stackName}_FORWARDER_ADDRESS`;
+
+const ethers = require('ethers');
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/autotasks/relay/index.js:4]:
```suggestion
const { ethers } = require('ethers');
```

File: defender/gasless-nft-minting/package.json [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  "author": "",
+  "license": "ISC",
+  "devDependencies": {
+    "chai": "4.3.7",
+    "chai-as-promised": "7.1.1",
+    "defender-serverless": "1.0.4",
+    "eslint": "8.34.0",
+    "eslint-config-airbnb-base": "15.0.0",
+    "eslint-plugin-import": "2.27.5",
+    "serverless": "^3.20.0"
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/package.json:22]:
```suggestion
    "serverless": "3.20.0"
```

File: defender/gasless-nft-minting/serverless.yml [hunk 12]
```diff
@@ -0,0 +1,41 @@
+service: governance-alert
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/serverless.yml:1]:
```suggestion
service: gasless-nft-minter
```

File: defender/gasless-nft-minting/tests/eip712sig.online.test.js [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+    const { request, signature } = await signMetaTxRequest(liveSigner, liveForwarderContract, input);
+
+    // Verify it
+    expect(await liveForwarderContract.verify(request, signature)).to.equal(true);
+  });
+
+
+  it('OZ Relayer can sign a message with EIP-151 standard', async function () {
+    // Sign a with the relayer
+    const message = 'Funds are safu!';
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/tests/eip712sig.online.test.js:112]: 🥇

File: defender/gasless-nft-minting/README.md [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+- Click on the **Relay** link on the left side
+- Find the Relay that deployed the NFT contract, click on the **copy address** button
+- Transfer funds to that address to cover the deployment and NFT minting costs
+
+### Local Code Setup
+
+In the `gasless-nft-minting` directory, perform the following steps:
+
+- A secrets file can be created for each stage of production. We will create one for development:
+  - Copy and rename the `sample.secrets.yml` to `.secrets/dev.yml`
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/README.md:54]: The `sample.secrets.yml` file is not in the `gasless-nft-minting` directory. It is in the `defender` directory.

File: defender/gasless-nft-minting/README.md [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+- Take the address of the `MinimalForwarder` and provide it as a constructor for the NFT contract
+  - `yarn deploy contract --contract-name DemoNFT "0xMINIMAL_FORWARDER_ADDRESS"`
+- Make a note of both newly deployed contract addresses. They will need to be added to the config file mentioned below
+
+### Serverless Deployment Setup :rocket:
+
+Now that the contracts are deployed, they can be added to the config file so that the Autotasks know which contracts they are signing for and forwarding to. In the `gasless-nft-minting` directory, perform the following steps:
+
+- A config file can be created for each stage of production. We will create one for development
+  - Copy and rename `sample.config.yml` to `config.dev.yml`
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/README.md:83]:
```suggestion
  - Copy and rename `example.config.yml` to `config.dev.yml`
```

File: defender/gasless-nft-minting/README.md [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+- Connect Relays to deployed Autotasks
+  - Contract Deployer relay must be connected to the signing Autotask
+  - Any Relayer with funds can be connected to the Relay Autotask
+- Copy the 2 Autotask webhooks into the `config.dev.yml` file
+
+### Defender Account Setup
+
+- In your [Defender account](https://defender.openzeppelin.com/), select the Hamburger icon in the upper right corner and click on **Team API Keys**
+- In the Team API Keys screen, click **Create API Key**
+- Make sure that the options for **Manage Autotasks** and **Manage Proposals and Contracts** are selected (we do not make use of the **Manage Relayers** and **Manage Sentinels** options with this template)
```
Comment (arbitrary-echo) [inline defender/gasless-nft-minting/README.md:24]: When I only selected "Manage Autotasks" and "Manage Proposals and Contracts" I could not successfully deploy with Serverless. Here is the error output I received:
```shell
❯ serverless deploy --stage dev
Running "serverless" from node_modules
(node:66347) NOTE: We are formalizing our plans to enter AWS SDK for JavaScript (v2) into maintenance mode in 2023.

Please migrate your code to use AWS SDK for JavaScript (v3).
For more information, check the migration guide at https://a.co/7PzMCcy
(Use `node --trace-warnings ...` to show where the warning was created)
========================================================
Secrets
✔ Updated Secret (gasless_nft_minting_part_deux_FORWARDER_ADDRESS)
✔ Updated Secret (gasless_nft_minting_part_deux_NFT_ADDRESS)
Contracts
Relayers
Autotasks
✖ User is not authorized to access this resource
✖ User is not authorized to access this resource
Notifications
Environment: darwin, node 18.12.1, framework 3.27.0 (local) 3.27.0v (global), plugin 6.2.3, SDK 4.3.2
Docs:        docs.serverless.com
Support:     forum.serverless.com
Bugs:        github.com/serverless/serverless/issues

Error:
DefenderApiResponseError: Request failed with status code 403
    at rejectWithDefenderApiError (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/defender-base-client/lib/api/api.js:11:27)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/defender-sentinel-client/lib/api/index.js:85:20
    at async SentinelClient.apiCall (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/defender-base-client/lib/api/client.js:26:20)
    at async DefenderDeploy.deployCategories (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/defender-serverless/lib/cmd/deploy.js:381:31)
    at async DefenderDeploy.deploy (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/defender-serverless/lib/cmd/deploy.js:775:9)
    at async DefenderDeploy.requestConfirmation (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/defender-serverless/lib/cmd/deploy.js:131:9)
    at async PluginManager.runHooks (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/serverless/lib/classes/plugin-manager.js:530:9)
    at async PluginManager.invoke (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/serverless/lib/classes/plugin-manager.js:564:9)
    at async PluginManager.run (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/serverless/lib/classes/plugin-manager.js:604:7)
    at async Serverless.run (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/serverless/lib/serverless.js:179:5)
    at async /Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/serverless/scripts/serverless.js:812:9
```

When I created an API key that allowed me to manage all of the Defender resources, the deployment worked.

**Review by arbitrary-echo — CHANGES_REQUESTED** [review]
The directions for setting up the API key may need to be changed to include more permissions. See the comment on the README.md.

File: defender/gasless-nft-minting/README.md (hunk 14 above)
Comment (arbitrary-nomad) [reply defender/gasless-nft-minting/README.md:54]: Fixed

Comment (arbitrary-nomad) [reply defender/gasless-nft-minting/README.md:54]: 9daa9144369044130f497f8c4a41219aab381ec9

File: defender/gasless-nft-minting/README.md (hunk 15 above)
Comment (arbitrary-nomad) [reply defender/gasless-nft-minting/README.md:83]: renamed the file. 9daa9144369044130f497f8c4a41219aab381ec9

File: defender/gasless-nft-minting/README.md (hunk 16 above)
Comment (arbitrary-nomad) [reply defender/gasless-nft-minting/README.md]: Changed it to add all permissions 9daa9144369044130f497f8c4a41219aab381ec9

**Review by arbitrary-echo — APPROVED** [review]

File: defender/gasless-nft-minting/serverless.yml [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+
+custom:
+  config: ${file(./config.${self:provider.stage}.yml)}
+  secrets: ${file(../.secrets/${self:provider.stage}.yml)}
+
+defender:
+  key: ${self:custom.secrets.keys.defender-api-key}
+  secret: ${self:custom.secrets.keys.defender-api-secret}
+
+functions:
```
Comment (ggonzalez94) [inline defender/gasless-nft-minting/serverless.yml:19]: Wouldn't it be better if a relayer is also deployed and hooked into the autotasks? That wat whoever is deploying this only needs to fund that new relayer. They always have the option to switch manually to an already funded relayer

**Review by ggonzalez94 — CHANGES_REQUESTED** [review]
This looks good! Just left a few minor comments. Besides that I have a couple more general feedback/concerns:

- Readme readability: The readme is a bit hard to follow, since you have to jump between sections to set things up and some instructions appear in more than one section. Do you think we can try to make it more streamlined?
- In the near future having self contained templates(with it's own package.json and hardhat.config.js) is what makes more sense, but currently the repo is set up with a main hardhat.config and folders for contracts, scripts, etc. My concern is that this mix of apporaches will make it more confusing for people trying to use the templates. Any thoughts on that?

File: defender/gasless-nft-minting/serverless.yml (hunk 17 above)
Comment (arbitrary-nomad) [reply defender/gasless-nft-minting/serverless.yml:19]: It's a chicken and the egg problem. 
- The contract should be deployed by a relayer and the Autotask needs the deployed contract's address during deployment
- If you deploy the contract first, the stack relay won't exist yet
- If you deploy the stack first, then the contracts wort exist yet

This template is based on the Contract Wizard Deployer, which assumes that you might already have a funded relayer that can deploy contracts. For simplicity’s sake, the choice was made to not create an additional relay and instead connect the one that deployed the contracts initially

File: defender/gasless-nft-minting/autotasks/signer/index.js [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+  'function execute(tuple(address from, address to, uint256 value, uint256 gas, uint256 nonce, bytes data) req, bytes signature) payable returns (bool, bytes)',
+  'function getNonce(address from) view returns (uint256)',
+  'function verify(tuple(address from, address to, uint256 value, uint256 gas, uint256 nonce, bytes data) req, bytes signature) view returns (bool)'
+];
+
+async function handler(event) {
+  // Parse webhook payload
+  if (!event?.request?.body?.address) throw new Error(`address missing`);
+  const { address: recipientAddress } = event.request.body;
+  ethers.utils.getAddress(recipientAddress);
```
[the remark this answers is not in the document]
Comment (arbitrary-nomad) [reply defender/gasless-nft-minting/autotasks/signer/index.js:23]: Done 2f49a15eab9f29f2e03a42e8696715e43d5a3ecd

File: defender/gasless-nft-minting/serverless.yml (hunk 17 above)
Comment (ggonzalez94) [reply defender/gasless-nft-minting/serverless.yml:19]: ok, makes sense. Thanks for the detailed explanation!

**Comment by arbitrary-nomad** [comment]
@ggonzalez94 

The README is written in the order that the steps should be followed. Many things are dependent on other things, so jumping in the README is not advised.

Order of operations:
- Acquire Defender and Relay keys
- Add funds to the Relay 
- Deploy 2 contracts using the above funded relayer
- Update `config.dev.yml` with the above deployed contract addresses
- Deploy the Autotasks with `serverless deploy` with the above updated config file
- Connect the deployer relay to the new Autotasks because the initial deployer has ownership of the NFT contract
- Copy the 2 Autotask webhooks into the `config.dev.yml` file so that hardhat signing and relaying scripts can be used.

RE: Hardhat usage in a sub folder: The root hardhat has its own scripts and output folders. To use hardhat and not break anything at the root level, the simplest approach was to point hardhat at a custom hardhat config file. Also the root hardhat is written in TS and our scripts are in JS making it a little more complicated as well.

**Comment by ggonzalez94** [comment]
@arbitrary-nomad I actually meant that the Readme currently makes you jump between sections and is a bit hard to setup(for me at least it was). Take as an example the first section *Setup*, when you see the commands to do `yarn deploy` and to update `config.dev.yml` you try to do it right away, but the steps for doing so are in latter sections. I thinks some small formatting of the Readme could help a lot, since the content is great and has all the steps.

> RE: Hardhat usage in a sub folder: The root hardhat has its own scripts and output folders. To use hardhat and not break anything at the root level, the simplest approach was to point hardhat at a custom hardhat config file. Also the root hardhat is written in TS and our scripts are in JS making it a little more complicated as well.

If it's too much of a pain to refactor you don't have to do it, but just FYI you can use the `hardhat.config.ts` version at the root of the repo and include `JS` files in the scripts forlder

File: defender/gasless-nft-minting/README.md [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+- Copy your Ethereum Address, API key, and Secret key to a local file (you will **NOT** be able to view your Secret key again after this message box goes away)
+- Make sure that you really did copy your API key and Secret key to a local file
+- Check the box for **I’ve written down the secret key** and select **Close**
+
+### Add Funds to the Relay
+
+- In your [Defender account](https://defender.openzeppelin.com/)
+- Click on the **Relay** link on the left side
+- Find the Relay that deployed the NFT contract, click on the **copy address** button
+- Transfer funds to that address to cover the deployment and NFT minting costs
```
Comment (darian-ae) [inline defender/gasless-nft-minting/README.md:47]: I think this could be a bit misleading. At this point, it's saying to find the Relay that has already deployed the NFT contract, but then saying to transfer funds to the Relay address to cover the cost of deployment.

You could replace line 46 with something like `Find the Relay that will be used to deploy the NFT contract`

**Review by darian-ae — APPROVED** [review]

File: defender/gasless-nft-minting/README.md (hunk 19 above)
Comment (arbitrary-nomad) [reply defender/gasless-nft-minting/README.md]: Thanks, it has been updated.

**Comment by arbitrary-nomad** [comment]
@ggonzalez94 , thank you for clarifying the README issues and for the info on hardhat. I've updated the README and broke it down to 3 main sections of:
- Deploying Contracts
- Deploying Autotasks
- Signing and Relaying Requests (either with scripts or curl)

Unfortunately, the config file needs to be updated in each section with the results of the previous. I haven't figured out to simplify it any further. I hope this is easier to follow.

**Comment by ggonzalez94** [comment]
Thanks @arbitrary-nomad! This looks good. I just updated the readme to remove an incomplete sentence and moved the warning to the top to make sure that people are aware that this is for development purposes only and not intended for production usage directly.
Merging this now

**Review by ggonzalez94 — APPROVED** [review]
